File: iosApp/iosApp/Modules/Degen/DegenInteractor.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

protocol DegenInteractorLister: AnyObject {
    func handle(networkEvent event: NetworksEvent)
}

protocol DegenInteractor: AnyObject {
    var categoriesActive: [DAppCategory] { get }
    var categoriesInactive: [DAppCategory] { get }
    func addListener(_ listener: DegenInteractorLister)
    func removeListener(_ listener: DegenInteractorLister)
}

final class DefaultDegenInteractor {
    private let degenService: DegenService
    private let networksService: NetworksService
    
    private var listeners: [WeakContainer] = []

    init(
        _ degenService: DegenService,
        networksService: NetworksService
    ) {
        self.degenService = degenService
        self.networksService = networksService
    }
    
    deinit {
        print("[DEBUG][Interactor] deinit \(String(describing: self))")
    }
}

extension DefaultDegenInteractor: DegenInteractor {

    var categoriesActive: [DAppCategory] {
        degenService.categoriesActive
    }

    var categoriesInactive: [DAppCategory] {
        degenService.categoriesInactive
    }
    
    func addListener(_ listener: DegenInteractorLister) {
        listeners = [WeakContainer(listener)]
        networksService.add(listener__: self)
    }

    func removeListener(_ listener: DegenInteractorLister) {
        listeners = []
        networksService.remove(listener__: self)
    }
}

extension DefaultDegenInteractor: NetworksListener {

    func handle(event_: NetworksEvent) {
        emit(event_)
    }
}

private extension DefaultDegenInteractor {
    
    func emit(_ event: NetworksEvent) {
        listeners.forEach { $0.value?.handle(networkEvent: event) }
    }

    class WeakContainer {
        weak var value: DegenInteractorLister?

        init(_ value: DegenInteractorLister) {
            self.value = value
        }
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/degen/DegenInteractor.kt
package com.sonsofcrypto.web3walletcore.modules.degen

import com.sonsofcrypto.web3lib.services.networks.NetworksEvent
import com.sonsofcrypto.web3lib.services.networks.NetworksListener
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3walletcore.services.degen.DAppCategory
import com.sonsofcrypto.web3walletcore.services.degen.DegenService

interface DegenInteractorLister {
    fun handle(event: NetworksEvent)
}

interface DegenInteractor {
    val categoriesActive: List<DAppCategory>
    val categoriesInactive: List<DAppCategory>
    fun isVoidSigner(): Boolean
    fun add(listener: DegenInteractorLister)
    fun remove(listener: DegenInteractorLister)
}

class DefaultDegenInteractor(
    private val degenService: DegenService,
    private val networksService: NetworksService,
    private val walletService: WalletService,
): DegenInteractor, NetworksListener {
    private var listener: DegenInteractorLister? = null

    override val categoriesActive: List<DAppCategory> get() = degenService.categoriesActive()

    override val categoriesInactive: List<DAppCategory> get() = degenService.categoriesInactive()

    override fun isVoidSigner(): Boolean =
        walletService.isSelectedVoidSigner()

    override fun add(listener: DegenInteractorLister) {
        if (this.listener != null) { networksService.remove(this) }
        this.listener = listener
        networksService.add(this)
    }

    override fun remove(listener: DegenInteractorLister) {
        this.listener = null
        networksService.remove(this)
    }

    override fun handle(event: NetworksEvent) {
        listener?.handle(event)
    }
}
